### ISSUE TITLE
PartOf= does not propagate restart to units in activating state

### ISSUE THREAD

**Issue by dechamps** [systemd/systemd#13841]
**systemd version the issue has been seen with**
242-7

**Used distribution**
Debian Unstable

**Expected behaviour you didn't see**
When calling `systemctl restart` on a unit, units with a `PartOf=` dependency should be restarted, including those in *activating (start)* state.

**Unexpected behaviour you saw**
Units with a `PartOf=` dependency that are in *activating (start)* state are not restarted.

**Steps to reproduce the problem**

Create a dummy `test-a.target` unit.

Create a `test-b.service` unit with the following:
```
[Unit]
PartOf=test-a.target
[Service]
Type=oneshot
ExecStart=/bin/sleep 10
```

Run:

```shell
systemctl start --no-block test-b.service
sleep 3
# test-b.service is in activating (start) state
systemctl restart test-a.target
```

You will notice the third command does not restart `test-b.service`, despite the `PartOf=` dependency.

**Additional information**

This behaviour is surprising because `systemctl stop test-a.target` does propagate correctly, and `systemctl restart test-b.service` does work, even if `test-b.service` is in activating state.

**Workaround**

One workaround is `systemctl stop test-a.target; systemctl start test-a.target`, but that only works if `test-a.target` has a dependency on `test-b.service`.

**Comment by dechamps** [comment]
Just played around with `--show-transaction` and found the following:

When `test-b.service` is in activating state:
```
# systemctl restart --show-transaction test-a.target 
Enqueued anchor job 8184658 test-a.target/restart.
```

When `test-b.service` is in active state, where restart does propagate correctly:
```
# systemctl restart --show-transaction test-a.target
Enqueued anchor job 8185283 test-a.target/restart.
Enqueued auxiliary job 8185285 test-b.service/restart.
```

It looks like the restart job is not being enqueued when the dependent unit is in activating state - it's only enqueued if it is in active state.

**Comment by dechamps** [comment]
@keur you might be interested in looking into this, since you took care of #13850 and this might be a problem in or around the same code (thanks, by the way!)

**Comment by dechamps** [comment]
@keur you might be interested in looking into this, since you took care of #13850 and this might be a problem in or around the same code (thanks, by the way!)

**Comment by keur** [comment]
@dechamps Yeah, this is similar in that the job is being garbage collected as a redundant. Thing is, the activating state (at least to me) implies that the unit has not left the restart state, so what is being flagged as redundant is, in fact, redundant. The example you give here is a bit contrived since you can just remove `Type=oneshot` and get the expected behavior.

**Comment by dechamps** [comment]
@keur The main reason why this behaviour feels like a bug to me is because it's not consistent with the way `systemctl restart` works when operating on a unit directly. If I use `systemctl restart`, the units explicitly listed will be restarted even if they are in activating state. But that doesn't happen if it's done indirectly through `PartOf=`. That doesn't make a ton of sense to me.

> the activating state (at least to me) implies that the unit has not left the restart state, so what is being flagged as redundant is, in fact, redundant

I disagree. If a unit is stuck in activating state (which can happen for a variety of reasons, e.g. a stuck `ExecStartPre=` command, or a stuck oneshot service), I'd expect a restart to unstuck it.

In general, if I issue multiple restarts in quick succession, I don't want any of them to be ignored. (And especially I don't want the last one to be ignored.)

> The example you give here is a bit contrived since you can just remove `Type=oneshot` and get the expected behavior.

The example is just a minimal case to reproduce. The actual use case in which this behaviour is problematic is one in which a script calls `systemctl restart` to restart a bunch of services, grouped through a `PartOf=` dependency on an abstract target. The reason why that script wants to restart the units is because something "changed" (e.g. some configuration), and they want these units to restart so that they can pick up the change. In this context, it is imperative that, *every time* `systemctl restart` is called, *all `PartOf=`-chained units restart from scratch*, otherwise they might miss the change. That is true even if they are in activating state, because some of these units (typically oneshot units) assess the new state during activation, not after they started running.

**Comment by pdraganov1** [comment]
The behaviour is the same for inactive (dead) and failed states. I have a service which has RuntimeDirectory= the same as home directory of the User= which the service is started with. I need .pgpass file copied and chowned in that home directory after the service is started. I created another oneshot service with Required= and PartOf= which executes a simple bash script with cp and chown in it but it is not started again after the first service is restarted. I also tried to make it Type=simple, so the service is in failed state, but again it is not restarted.
